### DISCUSSION TITLE
ESPNow deepsleep and i2c

### DISCUSSION THREAD

**Discussion by uli99** [micropython/micropython#13349] [ESP32]
Hi,

first post here and a beginner at micropython...

I'm trying to use espnow to send HTU21 sensor data (ESP32-WROOM-Dev board) to a peer (also a  ESP32-WROOM-Dev board) and at startup the HTU21 data arrives as expected on the peer side. But after going to deep sleep and coming back the i2c communication seems to have trouble to talk with the HTU21 sensor, the espnow communication works after the deepsleep (I can see a text message arriving at the peer), but not the i2c value:

HTU21 Sensor side:

>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
Temperature: 22.63°C

Going to sleep...
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
Traceback (most recent call last):
  File "main.py", line 27, in <module>
  File "/lib/micropython_htu21df/htu21df.py", line 78, in temperature
OSError: [Errno 19] ENODEV
MicroPython v1.22.0 on 2023-12-27; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> 

Peer:
MPY: soft reboot
b'\x0c\xb8\x15\xc3\x88\x08' b'Starting...'
b'\x0c\xb8\x15\xc3\x88\x08' b'HTU21d Temp: 22.63'
b'\x0c\xb8\x15\xc3\x88\x08' b'Starting...'

**Receiver side code:**
```python
import` network
import espnow

sta = network.WLAN(network.STA_IF)
sta.active(True)
sta.disconnect()   # Because ESP8266 auto-connects to last Access Point

e = espnow.ESPNow()
e.active(True)

while True:
    host, msg = e.recv()
    if msg:             # msg == None if timeout in recv()
        print(host, msg)
        if msg == b'end':
            break
```

**Sensor-Code:**

```python
import network
import espnow
import time
import machine
import wifi
from machine import Pin, I2C
from micropython_htu21df import htu21df


i2c = I2C(0, sda=Pin(21), scl=Pin(22))  # Correct I2C pins for ESP32-WROOM
htu = htu21df.HTU21DF(i2c)

sta, ap = wifi.reset()  # DeepSleep: Reset wifi to AP off, STA on and disconnected

sta = network.WLAN(network.STA_IF)  # Or network.AP_IF
sta.config(channel=1)           # Wifi loses config after lightsleep()
sta.active(True)
sta.disconnect()      # For ESP8266

e = espnow.ESPNow()
e.active(True)

peer = b'\x0c\xb8\x15\xd6\x2d\x10'   # MAC address of peer's wifi interface
e.add_peer(peer)      # Must add_peer() before send()

e.send(peer, "Starting...")
print(f"Temperature: {htu.temperature:.2f}°C")
print("")
e.send(peer, f"HTU21d Temp: {htu.temperature:.2f}")
e.active(False)
sta.active(False)               # Disable the wifi before sleep
print('Going to sleep...')
machine.deepsleep(10000)          # Sleep for 10 seconds then reboot
```

On both ESP32-WROOM boards is software version ESP32_GENERIC-20231227-v1.22.0.bin installed.

Any help would be appreciated

Uli

**Comment by robert-hh** [comment]
Please enclose code into lines with three backticks ```
That keeps indentation.  If you use ```python
for the first line, there is syntax-highlighting as well.

**Reply by davefes** [reply]
A bit of code I often find useful (placed at the start of the application) is:
```py
# during development
#if machine.reset_cause() == machine.SOFT_RESET:
#    print('you got a SOFT_RESET, so do a machine.reset()')
#    sleep_ms(10)
#    machine.reset()
```
Try changing to `machine.DEEPSLEEP_RESET`

**Comment by uli99** [comment]
ok, I added that and now it does a machine.reset() twice but still it is not able to access the i2c slave again after that.
I still get an error as soon as the HTU21df library tries to write stuff to the i2c slave:

MPY: soft reboot
Temperature: 23.23°C

Going to sleep...
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
you got a DEEPSLEEP_RESET, so do a machine.reset()
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
Traceback (most recent call last):
  File "main.py", line 36, in <module>
  File "/lib/micropython_htu21df/htu21df.py", line 78, in temperature
OSError: [Errno 19] ENODEV
MicroPython v1.22.0 on 2023-12-27; Generic ESP32 module with ESP32
Type "help()" for more information.
>>>

i2c must be somehow messed up after deep sleep...

How can I rectify the mess if even a hardware reset will not do it?

**Reply by robert-hh** [reply]
There was a PR suggesting a I2C clear as specified. That's done by keeping SDA high and sending >= 9 clock pulses at SCL. You can do that with the Pin methods before instantiating the I2C object. See PR #13281.

**Comment by uli99** [comment]
Ok, I tested a little more and it does not seem to be a problem of deep sleep.  Even when I just press the reset button on the ESP32 board, reading the HTU21 fails, only when I start the program from Thonny by clicking the play button (run the current script) the program reads the HTU21, very strange.

**Reply by robert-hh** [reply]
Are you sure that you copied the script that is executed by Thonny to the ESP board? Thonny runs the version of the script that is on the PC.

**Reply by uli99** [reply]
yes I'm sure, when I change the print command it shows up on the peer side.

But yes the reason for the not properly working program is most likely the user in front of the monitor =:)

I'll try another HTU21 library and see what will happen

**Comment by uli99** [comment]
Ok, I found this https://forum.micropython.org/viewtopic.php?t=12508
and I inserted a i2c.scan()

```python
i2c = I2C(0, sda=Pin(21), scl=Pin(22), freq=100000)  # Correct I2C pins for ESP32-WROOM
i2c.scan()
htu = htu21df.HTU21DF(i2c)
```
and now it works after a hard reset and a deep sleep reset...

I'll try the tip with the i2c clear robert-hh suggested tomorrow, that would be more efficient if it works, because scanning the whole i2c bus consumes certainly more time (energy) than >=9 clock cycles and a SDA high in  a deep sleep scenario.

**Reply by robert-hh** [reply]
That sounds good. I2C.scan() may do effectively the same as the clear_bus operation I mentioned above. But just calling I2C.scan() is simpler.

**Comment by uli99** [comment]
Ok, I added the i2c_clear function and this seems to do the trick as well:

```python
import network
import espnow
import time
import machine
import wifi
from machine import Pin, I2C
from micropython_htu21df import htu21df


def i2c_clear():
    SCL = Pin(22, Pin.OUT)
    SDA = Pin(21, Pin.OUT)
    SDA.value(0)
    time.sleep_us(2)
    SDA.value(1)
    for n in range(10):
        SCL.value(0)
        time.sleep_us(2)
        SCL.value(1)
        time.sleep_us(2)
        SCL.value(0)
    SDA.value(0)

i2c = I2C(0, sda=Pin(21), scl=Pin(22), freq=400000)  # Correct I2C pins for ESP32-WROOM
# i2c.scan()   # this is needed to fix error message after hardware reset...
i2c_clear()
htu = htu21df.HTU21DF(i2c)

sta, ap = wifi.reset()  # DeepSleep: Reset wifi to AP off, STA on and disconnected
# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF)  # Or network.AP_IF
sta.config(channel=1)           # Wifi loses config after lightsleep()
sta.active(True)
sta.disconnect()      # For ESP8266

e = espnow.ESPNow()
e.active(True)

peer = b'\x0c\xb8\x15\xd6\x2d\x10'   # MAC address of peer's wifi interface
e.add_peer(peer)      # Must add_peer() before send()

e.send(peer, "Starting...test")
# print(f"Temperature: {htu.temperature:.2f}°C")
print(f"Humidity: {htu.humidity:.2f}%")
print("")
# e.send(peer, f"HTU21d Temp: {htu.temperature:.2f}")
e.send(peer, f"HTU21d Hum: {htu.humidity:.2f}")
# e.send(peer, "testmessage")
e.active(False)
sta.active(False)               # Disable the wifi before sleep
print('Going to sleep...')
machine.deepsleep(10000)          # Sleep for 10 seconds then reboot
```
![MicroPython_i2c_clear_s](https://github.com/micropython/micropython/assets/17300608/d63f16b6-2155-4701-a850-f4dca5d0a963)

This still is very odd, running just the HTU21 with i2c and deep sleep works without the clear, but if I include espnow it needs the clear function to work...

Thanks

Uli

**Reply by robert-hh** [reply]
Thank you for testing. I was about to verify that as well. The function i2c_clear could eventually be simplified a bit, but it does not really matter.
```
def i2c_clear():
    SCL = Pin(22, Pin.OPEN_DRAIN, value=1)
    SDA = Pin(21, Pin.OPEN_DRAIN, value=1)
    time.sleep_us(2)
    for n in range(10):
        SCL(0)
        time.sleep_us(2)
        SCL(1)
        time.sleep_us(2)
```

**Comment by uli99** [comment]
toggeling SDA to 0 seems to be necessary, for me only this works:

```python
def i2c_clear():
    SCL = Pin(22, Pin.OPEN_DRAIN, value=1)
    SDA = Pin(21, Pin.OPEN_DRAIN, value=1)
    SDA(0)
    time.sleep_us(2)
    SCL(1)
    for n in range(10):
        SCL(0)
        time.sleep_us(2)
        SCL(1)
        time.sleep_us(2)
```

Uli

**Reply by robert-hh** [reply]
That's interesting. Below is a copy from the I2C specification of NXP:
```
3.1.16 Bus clear
In the unlikely event where the clock (SCL) is stuck LOW, the preferential procedure is to
reset the bus using the HW reset signal if your I2C devices have HW reset inputs. If the
I2C devices do not have HW reset inputs, cycle power to the devices to activate the
mandatory internal Power-On Reset (POR) circuit.
If the data line (SDA) is stuck LOW, the master should send nine clock pulses. The device
that held the bus LOW should release it sometime within those nine clocks. If not, then
use the HW reset or cycle power to clear the bus.
```
n.b.: Are you sure that it is necessary to keep SDA low during the nine clock cycles? And the SCL(1) before the loop seems not to be required, since the constructor already sets SCL high.

**Reply by uli99** [reply]
I tried it without that and it didn't work, so yes I'm sure it fixes my problem but might not really be a correct i2c-clear as intended by NXP.

**Reply by uli99** [reply]
If you look at the logic trace you can see that the SCL  line was high (not stuck low) before I called the clear function, not the situation the clear was intended for, but somehow it fixes the problem...

**Reply by robert-hh** [reply]
I've seen that, thanks. The NXP text is about two situations. First about SCL stuck low. That cannot handled with this code. Second about SDA stuck low. That can be cleared with the given procedure. On your logic trace both SDA and SCL are high. Nevertheless, clocking SCL clears the sensor's communication.

> This still is very odd, running just the HTU21 with i2c and deep sleep works without the clear, but if I include espnow it needs the clear function to work...

Whatever ESPnow causes.
